core/worker.h

```diff
@@ -167,25 +135,26 @@ class Worker {
   uint64_t current_tsc_;
   uint64_t current_ns_;
 
-  /* The current input gate index is not given as a function parameter.
-   * Modules should use get_igate() for access */
-  gate_idx_t igate_stack_[MAX_MODULES_PER_PATH];
-  int stack_depth_;
-
   // Gates and packets that this worker should serve next.
   struct gate_task gate_servicing_stack_[MAX_MODULES_PER_PATH * BRANCH_FACTOR];
   int pending_gates_;
 
+  /* The current input gate index is not given as a function parameter.
+   * Modules should use get_igate() for access */
+  gate_idx_t current_igate_;
+
   /* better be the last field. it's huge */
   struct pkt_batch splits_[MAX_GATES + 1];
-
-  DISALLOW_COPY_AND_ASSIGN(Worker);
 };
 
 extern int num_workers;
 extern std::thread worker_threads[MAX_WORKERS];
 extern Worker *volatile workers[MAX_WORKERS];
-extern thread_local Worker ctx;
+
+// NOTE: Do not use "thread_local" here. It requires a function call every time
+// it is accessed. Use __thread instead, which incurs minimal runtime overhead.
+// Worker must be "trivial" type.
```

Comment:
might be worthwhile to static assert on std::is_trivial (http://en.cppreference.com/w/cpp/types/is_trivial) here